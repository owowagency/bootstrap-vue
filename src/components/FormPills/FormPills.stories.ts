import FormPills from '.';
import {ref} from 'vue';

export default {
    title: 'Forms/Form pills',
    component: FormPills,
    argTypes: {
        editItemOnBackspace: {
            control: {type: 'boolean'},
        },
        maxItems: {
            control: {type: 'number'},
        },
        placeholder: {
            control: {type: 'text'},
        },
        valueMatcher: {
            control: {type: 'text'},
        },
    },
};

const template = (args) => ({
    components: {FormPills},
    setup: () => ({args, value: ref([])}),
    template: '<FormPills v-model="value" v-bind="args" />',
});

export const Default = template.bind({});
