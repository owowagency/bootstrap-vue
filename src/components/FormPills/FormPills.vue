<template>
    <label
        class="form-control"
        :for="id"
    >
        <ul class="list-unstyled d-flex align-content-start flex-wrap mb-0">
            <li
                v-for="(item, i) in modelValue"
                :key="i"
                class="form-pill d-inline-flex align-items-center mw-100"
            >
                <div class="form-pill-text">
                    {{ item }}
                </div>

                <div
                    class="form-pill-remove"
                    @click="removeItem(i)"
                />
            </li>

            <li class="form-pill-input-item">
                <div class="form-pill-input-sizer overflow-hidden">
                    {{ value }}
                </div>

                <input
                    :id="id"
                    v-model="value"
                    class="form-pill-input w-100"
                    type="text"
                    :placeholder="items.length ? '' : placeholder"
                    @blur="blur($event)"
                    @keydown="keydown($event)"
                    @paste="paste($event)"
                >
            </li>
        </ul>
    </label>
</template>

<script lang="ts">
import {PropType, computed, ref, watch} from 'vue';
import {idProps} from '@/composables/useId';
</script>

<script lang="ts" setup>
const props = defineProps({
    editItemOnBackspace: {
        type: Boolean,
        default: false,
    },
    ...idProps,
    maxItems: {
        type: Number,
        default: Infinity,
    },
    modelValue: {
        type: Array as PropType<string[]>,
        default: () => [],
    },
    placeholder: {
        type: String,
        default: '',
    },
    separator: {
        type: [String, RegExp],
        // For some reason it does not allow RegExp to be used as default value.
        // eslint-disable-next-line vue/require-valid-default-prop
        default: /[\s\n]+/,
    },
    submitKeys: {
        type: Array as PropType<string[]>,
        default: () => ['Enter'],
    },
    valueMatcher: {
        type: [String, RegExp],
        // For some reason it does not allow RegExp to be used as default value.
        // eslint-disable-next-line vue/require-valid-default-prop
        default: /^.+$/,
    },
});

const emit = defineEmits(['update:modelValue']);

const value = ref('');

const clearValue = () => value.value = '';

const items = computed({
    get: () => props.modelValue,
    set: (v: string[]) => emit('update:modelValue', v),
});

const maxItemsReached = computed(() => items.value.length >= props.maxItems);

const matchItem = (item: string) => !!item.match(props.valueMatcher);

const splitItem = (item: string) => item.split(props.separator);

const removeItem = (index: number) => items.value.splice(index, 1);

const validateValue = (value: string|string[]) => {
    if (Array.isArray(value)) {
        return value.every(validateValue);
    }

    return matchItem(value);
};

const addItem = (item: string) => {
    if (maxItemsReached.value) {
        return false;
    }

    const newItems = splitItem(item).filter(matchItem);

    if (newItems.length === 0) {
        return false;
    }

    const allowedItems = newItems.slice(0, props.maxItems - items.value.length);

    items.value = items.value.concat(allowedItems);

    return true;
};

watch(value, (v) => {
    const splitItems = splitItem(v);

    if (
        splitItems.length > 1
        && splitItems.filter(validateValue).length === splitItems.filter(v => v !== '').length
    ) {
        addItem(v);

        clearValue();
    }
});

// We need to pass the Focus Event to the blur method in order for the test to
// be able to mock the blur method.
const blur = (event: FocusEvent) => {
    const splitItems = splitItem(value.value);

    if (validateValue(splitItems)) {
        addItem(value.value);

        clearValue();
    }
};

const keydown = (event: KeyboardEvent) => {
    if (event.code === 'Backspace') {
        if (value.value !== '') {
            return;
        }

        const [removedItem] = removeItem(items.value.length - 1);

        if (props.editItemOnBackspace && removedItem) {
            value.value = removedItem;

            event.preventDefault();
        }

        return;
    }

    // If the max allowed items has been reached, don't allow any more typing.
    if (maxItemsReached.value) {
        event.preventDefault();

        return;
    }

    if (props.submitKeys.includes(event.code)) {
        const splitItems = splitItem(value.value);

        if (validateValue(splitItems)) {
            addItem(value.value);

            clearValue();
        }

        event.preventDefault();
    }
};

const paste = (event: ClipboardEvent) => {
    event.preventDefault();

    const pasteValue = event.clipboardData.getData('text');

    if (!pasteValue) {
        return;
    }

    const newValue = value.value + pasteValue;

    const splitItems = splitItem(newValue);

    // If the input value together with the paste value are valid, add the items,
    // otherwise concat the paste value to the input value.
    if (validateValue(splitItems)) {
        addItem(newValue);

        clearValue();
    } else {
        value.value = newValue;
    }
};
</script>
