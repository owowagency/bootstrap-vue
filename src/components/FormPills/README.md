# FormPills

```vue
<FormPills v-model="emails" :value-matcher="/.+@.+\.\w+/" />
```

## Imported Props

| Prop name | Description                   | Type   | Values | Default |
| --------- | ----------------------------- | ------ | ------ | ------- |
| id        | The id used for the component | string | -      | `uuid`  |

## Props

| Prop name           | Description                                                                | Type           | Values | Default            |
| ------------------- | -------------------------------------------------------------------------- | -------------- | ------ | ------------------ |
| editItemOnBackspace | Defines if items should be removed when pressing backspace                 | boolean        | -      | false              |
| maxItems            | The maximym amount of items that can be added to the list                  | number         | -      | Infinity           |
| modelValue          | The model value                                                            | string[]       | -      | () =&gt; []        |
| placeholder         | The placeholder for the input                                              | string         | -      | ''                 |
| separator           | The separator that checks if an item should be separated from the next one | string\|RegExp | -      | /[\s\n]+/          |
| submitKeys          | The key to press to submit an item                                         | string[]       | -      | () =&gt; ['Enter'] |
| valueMatcher        | The matcher to check if an item is valid                                   | string\|RegExp | -      | /^.+$/             |

## Events

| Event name        | Properties                           | Description                           |
| ----------------- | ------------------------------------ | ------------------------------------- |
| update:modelValue | **value** `string[]` - The new value | Fired when the model value is updated |
