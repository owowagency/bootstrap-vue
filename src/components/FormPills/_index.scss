.form-pill {
    @extend .badge;
    @extend .bg-primary;

    margin: .125rem .5rem .125rem 0;
}

.form-pill-text {
    @extend .text-truncate;
}

.form-pill-remove {
    padding: $badge-padding-y $badge-padding-x;
    margin: calc($badge-padding-y * -1) calc($badge-padding-x * -1) calc($badge-padding-y * -1) 0;
    cursor: pointer;
}

.form-pill-remove::after {
    content: 'x';
}

.form-pill-input-item {
    max-width: 100%;
}

.form-pill-input {
    background: transparent;
    border-width: 0;
    color: $input-color;
    outline: 0;
    padding: 0;
}

.form-pill-input-sizer {
    height: 0;
}
