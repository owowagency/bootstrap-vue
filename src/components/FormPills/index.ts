import FormPills from './FormPills.vue';

export default FormPills;
