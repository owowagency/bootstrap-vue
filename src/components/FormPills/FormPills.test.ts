import {describe, expect, it, vi} from 'vitest';
import FormPills from '.';
import {nextTick} from 'vue';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(FormPills, {
        props: {
            // Force an id to prevent snapshots from failing because of uuid.
            id: 'dont-change-please',
            modelValue: ['aa', 'bb', 'cc'],
        },
    });

    it('removes an item on click', async() => {
        const wrapper = shallowMount(FormPills, {props: {modelValue: ['aa']}});

        wrapper.vm.removeItem = vi.fn();

        await wrapper.find('.form-pill-remove').trigger('click');

        expect(wrapper.vm.removeItem).toBeCalledWith(0);
    });

    it('tries to add item on blur', async() => {
        const wrapper = shallowMount(FormPills, {props: {modelValue: ['aa']}});

        wrapper.vm.blur = vi.fn();

        await wrapper.find('input').trigger('blur');

        expect(wrapper.vm.blur).toBeCalled();
    });

    it('tries to add item on keydown', async() => {
        const wrapper = shallowMount(FormPills, {props: {modelValue: ['aa']}});

        wrapper.vm.keydown = vi.fn();

        const event = {code: 'Enter'};

        await wrapper.find('input').trigger('keydown', event);

        expect(wrapper.vm.keydown).toBeCalledWith(
            expect.objectContaining(event),
        );
    });

    it('tries to add item on paste', async() => {
        const wrapper = shallowMount(FormPills, {props: {modelValue: ['aa']}});

        wrapper.vm.paste = vi.fn();

        const event = {clipboardData: {getData: () => 'something nsfw'}};

        await wrapper.find('input').trigger('paste', event);

        expect(wrapper.vm.paste).toBeCalledWith(
            expect.objectContaining(event),
        );
    });
});

describe('clearValue', () => {
    it('clears value', () => {
        const wrapper = shallowMount(FormPills);

        wrapper.vm.value = 'something';

        wrapper.vm.clearValue();

        expect(wrapper.vm.value).toBe('');
    });
});

describe('items', () => {
    it('equals to modelValue', () => {
        const modelValue = ['a'];

        const wrapper = shallowMount(FormPills, {props: {modelValue}});

        expect(wrapper.vm.items).toEqual(['a']);

        modelValue.push('b');

        expect(wrapper.vm.items).toEqual(['a', 'b']);
    });

    it('emits update:modelValue when set', () => {
        const modelValue = ['a'];

        const wrapper = shallowMount(FormPills, {props: {modelValue}});

        wrapper.vm.items = ['a', 'b'];

        expect(wrapper.emitted('update:modelValue')[0]).toEqual([['a', 'b']]);
    });
});

describe('matchItem', () => {
    it('matches an item', () => {
        const wrapper = shallowMount(FormPills);

        expect(wrapper.vm.matchItem('I am an item')).toBe(true);
    });

    it('does not match an invalid item', () => {
        const wrapper = shallowMount(FormPills);

        expect(wrapper.vm.matchItem('')).toBe(false);
    });
});

describe('maxItemsReached', () => {
    it('computes max items not reached', () => {
        const wrapper = shallowMount(FormPills, {props: {
            maxItems: 3,
            modelValue: ['first', 'second'],
        }});

        expect(wrapper.vm.maxItemsReached).toBe(false);
    });

    it('computes max items reached', () => {
        const wrapper = shallowMount(FormPills, {props: {
            maxItems: 3,
            modelValue: ['first', 'second', 'third'],
        }});

        expect(wrapper.vm.maxItemsReached).toBe(true);
    });
});

describe('splitItem', () => {
    it('splits spaces and newlines by default', () => {
        const wrapper = shallowMount(FormPills);

        expect(wrapper.vm.splitItem('a a\na')).toEqual(['a', 'a', 'a']);
    });

    it('splits with custom separator', () => {
        const wrapper = shallowMount(FormPills, {props: {separator: ','}});

        expect(wrapper.vm.splitItem('a,a')).toEqual(['a', 'a']);
    });
});

describe('removeItem', () => {
    it('removes item on index', () => {
        // Cannot mock splice of `items` computed property.
        const modelValue = [];

        modelValue.splice = vi.fn();

        const wrapper = shallowMount(FormPills, {props: {modelValue}});

        wrapper.vm.removeItem(1);

        expect(modelValue.splice).toBeCalledWith(1, 1);
    });

    it('edits removed item on backspace', async() => {
        const wrapper = shallowMount(FormPills, {props: {
            editItemOnBackspace: true,
            modelValue: ['aa'],
        }});

        expect(wrapper.vm.value).toBe('');

        const event = {code: 'Backspace'};

        await wrapper.find('input').trigger('keydown', event);

        expect(wrapper.vm.value).toBe('aa');

        expect(wrapper.vm.modelValue).toHaveLength(0);
    });
});

describe('validateValue', () => {
    it('validates single value', () => {
        const wrapper = shallowMount(FormPills);

        expect(wrapper.vm.validateValue('item')).toBe(true);
    });

    it('validates array of values', () => {
        const wrapper = shallowMount(FormPills);

        expect(wrapper.vm.validateValue(['first', 'second', 'third'])).toBe(true);
    });

    it('invalidates single faulty value', () => {
        const wrapper = shallowMount(FormPills, {props: {valueMatcher: '^match'}});

        expect(wrapper.vm.validateValue('nomatch')).toBe(false);
    });

    it('invalidates single faulty value in array of values', () => {
        const wrapper = shallowMount(FormPills, {props: {valueMatcher: '^match'}});

        expect(wrapper.vm.validateValue(['match', 'nomatch', 'match again'])).toBe(false);
    });
});

describe('addItem', () => {
    it('adds new items', () => {
        // Cannot mock concat of `items` computed property.
        const modelValue = [];

        modelValue.concat = vi.fn();

        const wrapper = shallowMount(FormPills, {props: {modelValue}});

        expect(wrapper.vm.addItem('a')).toBe(true);

        expect(modelValue.concat).toBeCalledWith(['a']);
    });

    it('does not add items when max items reached', () => {
        const wrapper = shallowMount(FormPills, {props: {
            maxItems: 2,
            modelValue: ['first', 'second'],
        }});

        expect(wrapper.vm.addItem('third')).toBe(false);
    });

    it('does not add new items when empty split', () => {
        const wrapper = shallowMount(FormPills);

        expect(wrapper.vm.addItem('')).toBe(false);
    });

    it('adds new items with custom value matcher', () => {
        // Cannot mock concat of `items` computed property.
        const modelValue = [];

        modelValue.concat = vi.fn();

        const wrapper = shallowMount(FormPills, {props: {
            modelValue,
            valueMatcher: '^match',
        }});

        expect(wrapper.vm.addItem('match nomatch')).toBe(true);

        expect(modelValue.concat).toBeCalledWith(['match']);
    });

    it('does not add new item with non-matching custom value matcher', () => {
        const wrapper = shallowMount(FormPills, {props: {valueMatcher: '^match'}});

        expect(wrapper.vm.addItem('nomatch')).toBe(false);
    });

    it('adds new items until max reached', () => {
        // Cannot mock concat of `items` computed property.
        const modelValue = [];

        modelValue.concat = vi.fn();

        const wrapper = shallowMount(FormPills, {props: {
            modelValue,
            maxItems: 2,
        }});

        expect(wrapper.vm.addItem('first second third')).toBe(true);

        expect(modelValue.concat).toBeCalledWith(['first', 'second']);
    });
});

describe('value', () => {
    describe('watch', () => {
        it('adds item and clears itself', async() => {
            const wrapper = shallowMount(FormPills);

            // TODO: Unable to mock child method calls. Applies on more than
            // this test.

            // Trigger the watcher.
            wrapper.vm.value = 'a ';

            await nextTick();

            expect(wrapper.vm.value).toBe('');
        });

        it('does not add item when no items to be added', async() => {
            const wrapper = shallowMount(FormPills);

            // Trigger the watcher.
            wrapper.vm.value = 'a';

            await nextTick();

            expect(wrapper.vm.value).toBe('a');
        });
    });
});

describe('paste', () => {
    it('adds items', () => {
        // Cannot mock concat of `items` computed property.
        const modelValue = [];

        modelValue.concat = vi.fn();

        const wrapper = shallowMount(FormPills, {props: {modelValue}});

        const event = {
            clipboardData: {getData: vi.fn().mockReturnValue('something nsfw')},
            preventDefault: vi.fn(),
        };

        wrapper.vm.paste(event);

        expect(event.clipboardData.getData).toBeCalledWith('text');

        expect(event.preventDefault).toBeCalled();

        expect(modelValue.concat).toBeCalledWith(['something', 'nsfw']);
    });

    it('does not add items', () => {
        // Cannot mock concat of `items` computed property.
        const modelValue = [];

        modelValue.concat = vi.fn();

        const wrapper = shallowMount(FormPills, {props: {modelValue}});

        const event = {
            clipboardData: {getData: vi.fn().mockReturnValue('')},
            preventDefault: vi.fn(),
        };

        wrapper.vm.paste(event);

        expect(modelValue.concat).not.toBeCalled();
    });

    it('adds clipboard to input if data is invalid', () => {
        // Cannot mock concat of `items` computed property.
        const modelValue = [];

        modelValue.concat = vi.fn();

        const wrapper = shallowMount(FormPills, {
            props: {
                modelValue,
                valueMatcher: /^abc$/,
            },
        });

        const event = {
            clipboardData: {getData: vi.fn().mockReturnValue('abcd abc abcd')},
            preventDefault: vi.fn(),
        };

        wrapper.vm.paste(event);

        expect(wrapper.vm.value).toBe(event.clipboardData.getData());
        expect(modelValue.concat).not.toBeCalled();
    });
});
