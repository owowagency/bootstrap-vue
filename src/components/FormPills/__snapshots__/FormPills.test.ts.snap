// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<label class=\\"form-control\\" for=\\"dont-change-please\\">
  <ul class=\\"list-unstyled d-flex align-content-start flex-wrap mb-0\\">
    <li class=\\"form-pill d-inline-flex align-items-center mw-100\\">
      <div class=\\"form-pill-text\\">aa</div>
      <div class=\\"form-pill-remove\\"></div>
    </li>
    <li class=\\"form-pill d-inline-flex align-items-center mw-100\\">
      <div class=\\"form-pill-text\\">bb</div>
      <div class=\\"form-pill-remove\\"></div>
    </li>
    <li class=\\"form-pill d-inline-flex align-items-center mw-100\\">
      <div class=\\"form-pill-text\\">cc</div>
      <div class=\\"form-pill-remove\\"></div>
    </li>
    <li class=\\"form-pill-input-item\\">
      <div class=\\"form-pill-input-sizer overflow-hidden\\"></div><input id=\\"dont-change-please\\" class=\\"form-pill-input w-100\\" type=\\"text\\" placeholder=\\"\\">
    </li>
  </ul>
</label>"
`;
